File: ti4j-core/src/main/java/com/dpf/ti4j/core/TypeInspector.java
package com.dpf.ti4j.core;

import java.math.BigDecimal;
import java.math.BigInteger;
import java.nio.file.FileSystems;
import java.nio.file.Path;
import java.time.*;
import java.util.*;

final class TypeInspector {

    private static final Class<?>[] KNOWN_IMMUTABLES = {
            // Add known immutable classes here
    };

    private static final Class<?>[] KNOWN_MUTABLES = {
            // Add known mutable classes here
    };

    private static final Class<?>[] JAVA_IMMUTABLES = {
            Integer.class,
            Long.class,
            Double.class,
            Boolean.class,
            Character.class,
            Byte.class,
            Short.class,
            Float.class,
            String.class,
            Instant.class,
            LocalDate.class,
            LocalTime.class,
            LocalDateTime.class,
            ZonedDateTime.class,
            OffsetDateTime.class,
            OffsetTime.class,
            Duration.class,
            Period.class,
            Optional.class,
            OptionalInt.class,
            OptionalLong.class,
            OptionalDouble.class,
            UUID.class,
            Path.class,
            FileSystems.class,
            BigInteger.class,
            BigDecimal.class,
    };

    private static final String[] IMMUTABLE_COLLECTIONS = {
            "java.util.Collections$Unmodifiable",
            "java.util.ImmutableCollections$"
    };

    private static final String[] IMMUTABLE_MAPS = {
            "java.util.Collections$Unmodifiable",
            "java.util.ImmutableCollections$"
    };

    static boolean isKnownMutable(Class<?> clazz) {
        return Arrays.asList(KNOWN_MUTABLES).contains(clazz);
    }

    static boolean isKnownImmutable(Class<?> clazz) {
        return Arrays.asList(KNOWN_IMMUTABLES).contains(clazz);
    }

    static boolean isJavaImmutable(Class<?> clazz) {
        return Arrays.asList(JAVA_IMMUTABLES).contains(clazz);
    }

    static boolean isImmutableCollection(Object obj) {
        if (obj == null) {
            return false;
        }
        final var className = obj.getClass().getName();
        return Arrays.stream(IMMUTABLE_COLLECTIONS)
                .anyMatch(className::startsWith);
    }

    static boolean isImmutableMap(Object obj) {
        if (obj == null) {
            return false;
        }
        final var className = obj.getClass().getName();
        return Arrays.stream(IMMUTABLE_MAPS)
                .anyMatch(className::startsWith);
    }
}

File: src/main/kotlin/com/dpf/ti4j/validation/TypeInspector.kt
package com.dpf.ti4j.validation

import java.math.BigDecimal
import java.math.BigInteger
import java.nio.file.FileSystems
import java.nio.file.Path
import java.time.Duration
import java.time.Instant
import java.time.LocalDate
import java.time.LocalDateTime
import java.time.LocalTime
import java.time.OffsetDateTime
import java.time.OffsetTime
import java.time.Period
import java.time.ZonedDateTime
import java.util.Optional
import java.util.OptionalDouble
import java.util.OptionalInt
import java.util.OptionalLong
import java.util.UUID
import java.util.concurrent.atomic.AtomicReference

object TypeInspector {

    fun isJavaImmutable(clazz: Class<*>): Boolean {
        return when (clazz) {
            Integer::class.java,
            java.lang.Long::class.java,
            java.lang.Double::class.java,
            java.lang.Boolean::class.java,
            Character::class.java,
            java.lang.Byte::class.java,
            java.lang.Short::class.java,
            java.lang.Float::class.java,
            String::class.java,
            Instant::class.java,
            LocalDate::class.java,
            LocalTime::class.java,
            LocalDateTime::class.java,
            ZonedDateTime::class.java,
            OffsetDateTime::class.java,
            OffsetTime::class.java,
            Duration::class.java,
            Period::class.java,
            Optional::class.java,
            OptionalInt::class.java,
            OptionalLong::class.java,
            OptionalDouble::class.java,
            UUID::class.java,
            Path::class.java,
            FileSystems::class.java,
            BigInteger::class.java,
            BigDecimal::class.java,
                -> true

            else -> false
        }
    }

    fun isAtomicReference(obj: Any?): Boolean {
        return obj is AtomicReference<*>
    }

    fun isImmutableCollection(obj: Any?): Boolean {
        return when (obj?.javaClass?.name) {
            "java.util.Collections\$UnmodifiableList",
            "java.util.Collections\$UnmodifiableRandomAccessList",
            "java.util.Collections\$UnmodifiableSet" -> true

            else -> obj?.javaClass?.name?.startsWith("java.util.ImmutableCollections") == true
        }
    }

    fun isImmutableMap(obj: Any?): Boolean {
        return when (obj?.javaClass?.name) {
            "java.util.Collections\$UnmodifiableMap" -> true
            else -> obj?.javaClass?.name?.startsWith("java.util.ImmutableCollections") == true
        }
    }
}